### PR TITLE
feat: store dir structure compatible with pnpm

### PR DESCRIPTION
**Pull request by KSXGitHub**
Resolves https://github.com/pnpm/pacquet/issues/165

**Other changes:**
* A new crate named `pacquet-store-dir` has been created.
* The field `store-dir` in `Npmrc` has been changed from `PathBuf` to `StoreDir`.
* The function `write_sync` in `pacquet-cafs` has been replaced by 2 functions (https://github.com/pnpm/pacquet/commit/8f98e730c1c35e2371b1d5ebd7f9e3f86b99104e).
* The `pacquet-cafs` crate has been merged into `pacquet-store-dir` (https://github.com/pnpm/pacquet/commit/db673efdc933a27a10eef3ae46b2f3f5870fcc96).
* The `pacquet-tarball` crate now also writes index files (3b6e68a48b0711a2666bbe4964b9fd485a9842d3..8f98e730c1c35e2371b1d5ebd7f9e3f86b99104e).
* The command `pacquet prune` now panics on `todo!()` for being incomplete (https://github.com/pnpm/pacquet/commit/c8526d0d33ad270bc499e65b7c48d0191ea47811). 

---

**TODO:**
* [x] Test:
  - [x] `cas_file_path`.
  - [x] `index_file_path`.
  - [x] Snapshot of the store.
  - [x] Use `pnpm install --offline` to test store dir compatibility.
  - [x] Executable bits.
* [x] Use `sha2::Sha512`.
* [x] Executable files (https://github.com/pnpm/pacquet/pull/166/commits/31148d3e802b3ef426b7d12e8e1efdeb67b9f0fc).
* [x] Fix executable bits
  * [x] Use 111 as bit mask.
  * [x] All executables must be 755.
* [x] Index files (3b6e68a48b0711a2666bbe4964b9fd485a9842d3..8f98e730c1c35e2371b1d5ebd7f9e3f86b99104e).
* [x] Disable `prune` (https://github.com/pnpm/pacquet/pull/166/commits/c8526d0d33ad270bc499e65b7c48d0191ea47811).
* [x] `checkedAt`.
* [x] Refactor:
  - [x] Remove unnecessary items from `cafs`.
  - [x] Create `StoreDir::new` and switch all `.to_path_buf().into()` to using it.
  - [x] Move `cafs` into `store-dir` (https://github.com/pnpm/pacquet/pull/166/commits/db673efdc933a27a10eef3ae46b2f3f5870fcc96).
  - [x] Add `From` to `TaskError`.
  - [x] Separate `write_sync` into 2 functions, one for non-index, one for index (https://github.com/pnpm/pacquet/pull/166/commits/8f98e730c1c35e2371b1d5ebd7f9e3f86b99104e).
  - [x] Move `TarballIndex` to `store-dir` and reuse it in `write_tarball_index_file`.
  - [x] `file_path_by_content_address` should take `exec: bool` instead of `suffix: FileSuffix`.
  - [x] Remove `FileSuffix` and `strum`.
  - [x] `pacquet_fs::make_file_executable`.

### CODE REVIEW

File: crates/store-dir/src/store_dir.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        root.into().into()
+    }
+
+    /// Create an object that [displays](std::fmt::Display) the root of the store directory.
+    pub fn display(&self) -> path::Display {
+        self.root.display()
+    }
+
+    /// Get `{store}/v3`.
+    fn v3(&self) -> PathBuf {
```
Comment (zkochan) [inline crates/store-dir/src/store_dir.rs:36]: maybe just call it currentVersion?
```suggestion
    fn currentVersion(&self) -> PathBuf {
```
or maybe root can be the v3 directory.

File: crates/tarball/src/lib.rs [hunk 2]
```diff
@@ -194,36 +209,75 @@ impl<'a> DownloadTarballToStore<'a> {
...
+
+                let tarball_index_key = cleaned_entry_path
+                    .to_str()
+                    .expect("entry path must be valid UTF-8") // TODO: propagate this error, provide more information
+                    .to_string(); // TODO: convert cleaned_entry_path to String too.
+
+                if let Some(previous) = cas_paths.insert(cleaned_entry_path, file_path) {
+                    panic!("Unexpected error: {previous:?} shouldn't collide");
+                }
+
```
Comment (zkochan) [inline crates/tarball/src/lib.rs:260]: This is not correct! The last occurrence of the file should be used. This is important. We have even a test in pnpm to verify this. We'll need to port the same test.

https://github.com/pnpm/pnpm/blob/4c2450208d15bb03797c8f4497321490d823d685/fetching/tarball-fetcher/test/fetch.ts#L468-L483

File: crates/tarball/src/lib.rs [hunk 3]
```diff
@@ -194,36 +209,75 @@ impl<'a> DownloadTarballToStore<'a> {
...
+                    checked_at,
+                    integrity: file_integrity,
+                    mode: file_mode,
+                    size: file_size,
+                };
+
+                if let Some(previous) = tarball_index.files.insert(tarball_index_key, file_attrs) {
+                    panic!("Unexpected error: {previous:?} shouldn't collide");
+                }
+            }
```
Comment (zkochan) [inline crates/tarball/src/lib.rs:274]: Same as above. The last file should be used.

**Review by zkochan — CHANGES_REQUESTED** [review]
The most critical comment is about the entries extraction. The last entry should be used when there are several entries with the same name.

File: crates/store-dir/src/index_file.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+use pacquet_fs::{ensure_file, EnsureFileError};
+use serde::{Deserialize, Serialize};
+use ssri::{Algorithm, Integrity};
+use std::{collections::HashMap, path::PathBuf};
+
+impl StoreDir {
+    /// Path to an index file of a tarball.
+    pub fn tarball_index_file_path(&self, tarball_integrity: &Integrity) -> PathBuf {
+        let (algorithm, hex) = tarball_integrity.to_hex();
+        assert_eq!(algorithm, Algorithm::Sha512, "Only Sha512 is supported"); // TODO: propagate this error
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/store-dir/src/index_file.rs:13]: How do we derive content address from sha1 algorithm?

File: crates/npmrc/src/custom_deserializer.rs [hunk 5]
```diff
@@ -67,13 +68,13 @@ pub fn default_store_dir() -> PathBuf {
...
-        return default_store_dir_windows(&home_dir, &current_dir);
+        return default_store_dir_windows(&home_dir, &current_dir).into();
     }
 
     // https://doc.rust-lang.org/std/env/consts/constant.OS.html
     match env::consts::OS {
-        "linux" => home_dir.join(".local/share/pacquet/store"),
-        "macos" => home_dir.join("Library/pacquet/store"),
+        "linux" => home_dir.join(".local/share/pacquet/store").into(),
+        "macos" => home_dir.join("Library/pacquet/store").into(),
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/npmrc/src/custom_deserializer.rs]: https://github.com/pnpm/pacquet/pull/166/commits/1fb2ec5985394e6f345d433985b86ff4aedfd1e3

[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/npmrc/src/custom_deserializer.rs]: https://github.com/pnpm/pacquet/pull/166/commits/27502716d904caeae5dc4d259884cdc53a593d08

File: crates/store-dir/src/store_dir.rs (hunk 1 above)
Comment (KSXGitHub) [reply crates/store-dir/src/store_dir.rs:36]: This functions returns the path to a sub dir named `v3`, not the version of the store.

File: crates/store-dir/src/index_file.rs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        self.file_path_by_hex_str(&hex, "-index.json")
+    }
+}
+
+/// Content of an index file (`$STORE_DIR/v3/files/*/*-index.json`).
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TarballIndex {
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/store-dir/src/index_file.rs]: https://github.com/pnpm/pacquet/pull/166/commits/4f8752294282aa5a1033a862ba62284a9a514215, https://github.com/pnpm/pacquet/pull/166/commits/3883db010dfdb4c101a7fed0d6f303317e7f000a

File: crates/store-dir/src/index_file.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    pub files: HashMap<String, TarballIndexFileAttrs>,
+}
+
+/// Value of the [`files`](TarballIndex::files) map.
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TarballIndexFileAttrs {
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/store-dir/src/index_file.rs]: https://github.com/pnpm/pacquet/pull/166/commits/2d54d459f349b9ef2ddc1d0340d79393aea644c1

File: crates/tarball/src/lib.rs (hunk 2 above)
Comment (KSXGitHub) [reply crates/tarball/src/lib.rs]: https://github.com/pnpm/pacquet/pull/166/commits/3ab34b855644f2ed22ecb87a35aa8a87922f1b9c

File: crates/tarball/src/lib.rs (hunk 3 above)
Comment (KSXGitHub) [reply crates/tarball/src/lib.rs]: https://github.com/pnpm/pacquet/pull/166/commits/3ab34b855644f2ed22ecb87a35aa8a87922f1b9c

File: crates/cli/tests/snapshots/install__should_install_exec_files.snap
Comment (zkochan) [inline crates/cli/tests/snapshots/install__should_install_exec_files.snap:1]: Why do we need so many files? Isn't it enough to test on a package that has two files? One executable and one non-executable?

**Comment by zkochan** [comment]
There's one important nuance to linking from the store. This shouldn't be done in this PR but we should remember to do it. When a package that has postinstall scripts is "imported" from the store, it is only imported using reflinks or copying. This is important because we build the package and during build the files could get modified. We don't want files in the store become modified.

Also, we will have to implement side-effects caching. But that's also for a separate PR and not high priority at the moment.

File: crates/cli/tests/snapshots/install__should_install_exec_files.snap
Comment (KSXGitHub) [reply crates/cli/tests/snapshots/install__should_install_exec_files.snap:1]: If you can suggest a package that has only a few files with one executable, I will use it.

File: crates/store-dir/src/index_file.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/store-dir/src/index_file.rs:13]: OK. But I still need to error on other algorithms.

[the remark this answers is not in the document]
Comment (KSXGitHub) [reply crates/store-dir/src/index_file.rs]: https://github.com/pnpm/pacquet/pull/166/commits/bbdfc2f32f249be5acb4525d13b4690212d1e5cb

File: crates/cli/tests/snapshots/install__should_install_exec_files.snap
Comment (KSXGitHub) [reply crates/cli/tests/snapshots/install__should_install_exec_files.snap:1]: There relevant code is here: https://github.com/pnpm/pacquet/blob/95741efac3224947de52e0800c42dde695d6fca5/crates/cli/tests/install.rs#L57-L61

**Review by zkochan — APPROVED** [review]

File: crates/cli/tests/snapshots/install__should_install_exec_files.snap
Comment (zkochan) [reply crates/cli/tests/snapshots/install__should_install_exec_files.snap:1]: We have a [registry-mock](https://github.com/pnpm/registry-mock) with testing packages. It has a package with one executable: https://github.com/pnpm/registry-mock/tree/main/packages/hello-world-js-bin%401.0.0

We just need to start using the registry-mock

Comment (KSXGitHub) [reply crates/cli/tests/snapshots/install__should_install_exec_files.snap:1]: > We just need to start using the registry-mock

This should be in another PR.

**Pull request merged**